### ISSUE TITLE
[html] XPath incorrectly flagging label/input pairs

### ISSUE THREAD

**Issue by codefriar** [pmd/pmd#4740]
**Affects PMD Version:** 
latest as of 11/7/23

**Rule:**

```xml
<rule name="xpathForLegends"
        language="html"
        message=""
        class="net.sourceforge.pmd.lang.rule.XPathRule">
        <description>
            
        </description>
        <priority>1</priority>
        <properties>
            <property name="version" value="3.1" />
            <property name="xpath">
                <value>
<![CDATA[
//*[
//fieldset[legend and not(.//input[not(@id or (descendant::label[@for = @id] or ancestor::label[@for = @id]))])]
]]>
                </value>
            </property>
        </properties>
    </rule>
```

**Description:**
this html snippet should *not* be flagged by the xpath rule, as all of it's input's have a corresponding label.
```html
<fieldset>
    <legend>test</legend>
     <label for="cheese">Do you like cheese?</label>
     <input type="checkbox" name="cheese" id="cheese" />
</fieldset>
```

**Expected outcome:**

This xpath should not flag this code snippet.

**Running PMD through:** CLI

**Comment by jsotuyod** [comment]
@codefriar I fear the issue is with your XPath, not with PMD…

Firstly, the XPath you copied is broken… there seems to be either an extra `//*[` or a missing `]` at the end.

Lastly, both `descendant::label[@for = @id]` and `ancestor::label[@for = @id]` will always fail for your sample HTML… you seem to want to match the `for` attribute of the `label` with the `id` of the `input`, but the way it's written it is looking for both attributes in `label`… as there is no `id` in `label`, it won't match; and since both expressions are within a `not` it will effectively match your `fieldset`.

I suggest you try debugging and testing your XPaths using [PMD's built-in designer](https://docs.pmd-code.org/latest/pmd_userdocs_extending_designer_reference.html)
